feat: Add config directory opt to CLI

asummers opened the pull request:
Resolves #16 

Note: I unfolded the `__MODULE__` alias expansion because I found grepping for where the `EXCheck.Config.*` modules were being used was tougher than it should have been, so I resolved this for future folks by making it possible to grep for the fully qualified name of the modules. I can put this back if you'd like, but I think it's a positive change. 

I tested this locally, but how would you like this tested for posterity?

karolsluszniak commented on lib/mix/tasks/check.ex:
```diff
@@ -165,13 +165,15 @@ defmodule Mix.Tasks.Check do
   1. Default stock configuration.
-  2. `.check.exs` in user home directory.
-  3. `.check.exs` in current project directory (or umbrella root for an umbrella project).
+  2. `.check.exs` in `--config-directory` opt on command line.
```
Personally I feel that it's quite unnatural to point to config directory instead of directly to a config file. I believe that most tools that have similar option allow to specify a file and not a dir. It also works more naturally with shell auto-completion. Aaand that'd be more in line with what you've originally suggested in #16, isn't it? What do you think?

Btw, coding that should not be that much more work - you'd have to change `load_from_dirs` to `load_from_files` and join `@config_filename` to directories in `load` instead of in `load_from_dirs`.

asummers replied:
Yeah I went with the directory since that was the tenor of the other code, not wanting to rock the boat too much. I agree that it pointing to the file seems more natural. Will change. :+1:

karolsluszniak commented on lib/ex_check/config.ex:
```diff
@@ -1,8 +1,6 @@
 defmodule ExCheck.Config do
   @moduledoc false
 
-  alias __MODULE__.{Generator, Loader}
```
I agree that using full module names has its benefits. On the other hand, there are more instances of `__MODULE__` that you haven't changed so that feels a little bit inconsistent. Up to you to decide :)

asummers replied:
I can change the others in a parallel PR. Will open it when I do the rest of this.

asummers replied:
Parallel PR opened in #22

karolsluszniak commented on lib/ex_check/check.ex:
```diff
@@ -5,7 +5,7 @@ defmodule ExCheck.Check do
 
   def run(opts) do
-    {tools, config_opts} = Config.load()
+    {tools, config_opts} = Config.load(opts)
```
Since after suggestion below you'll probably end up with `:config` option, I'd rename it to `:file` when passing to loader in order to keep the name relevant & specific in the context of config module.

asummers replied:
To be clear you want the CLI opt to remain `:config` and then it just gets `Keyword.replace` called on it to be `:file`? Or you would like the opt to be `:file` everywhere? 

Do we just want to name the opt `:config_file` and call it that everywhere?

karolsluszniak replied:
Since opts in the context of check module may include lots of keys irrelevant to config loading, I was thinking to only pass the relevant options and to name them with relevance to the config loader. Therefore, it'd be something like:

```
{tools, config_opts} = Config.load(file: opts[:config])
```

This way you aren't forced to invent a name that makes sense in all layers, like your `:config_file` that makes the CLI option unnecessarily long since `--config` is pretty self explanatory in that layer, but may be misleading in the context of config loader where `:config` could mean anything

What do you think? Would you rather keep your `config_file` or go for shorter `config` renamed to `file` when passed to the loader? :)

asummers replied:
I can rename.

asummers replied:
Okay take a look. The name shuffle from `:config` to `:file` got a little wordy in `Check.run1` because of the `Config.load/1` taking more opts, so I took opportunity to drop `:config` there as well. This should be closer to what you're thinking, I think.

karolsluszniak commented on lib/mix/tasks/check.ex:
```diff
@@ -165,13 +165,15 @@ defmodule Mix.Tasks.Check do
 
   ## Command line options
 
+  - `--config-directory /some/directory` - Override default config lookup directory
```
Please also advertise the change in `CHANGELOG.md`.

asummers commented:
Those changes all sound reasonable. Will change accordingly.

karolsluszniak commented:
@asummers Merged the others - now this is the last one standing after your pull requesting frenzy :) I'll wait a bit before releasing 0.13 to have this included.

asummers commented:
Ack. I will try to clean this PR up this week.

asummers commented:
I believe I have addressed all comments.

I just have one outstanding question in https://github.com/karolsluszniak/ex_check/pull/21#discussion_r513152444

I pushed a change to the opt being called `--config-file` everywhere, pursuant to my suggestion there.  I think it communicates the intent, but I am happy to rename if you feel that name is suboptimal.

asummers commented:
Sorry for the delays in getting back around on this. Happy to make any more changes and the turnaround should be much quicker =)
